implementation.go

```diff
@@ -0,0 +1,264 @@
+		return RowProviderAccessError{rr, err}
+	}
+	r.RetrievedValues = rr.RetrievedValues
+	css := make([]CacheAccess, len(rr.RetrievedValues), 0)
+	for _, rv := range rr.RetrievedValues {
+		key, err := f.Table.cacheKey(rv)
+		if err != nil {
+			return err
+		}
+		value, err := f.Table.cacheValue(rv)
+		if err != nil {
+			return err
+		}
+		cs := CacheSet{key.Key(), value.Value()}
+		css = append(css, cs)
+	}
+	errs := f.CacheProvider.AccessSlice(css)
+	nnerrs := []error(nil)
+	for i, err := range errs {
+		if err != nil {
+			if nnerrs == nil {
+				nnerrs = make([]error, len(errs), 0)
+			}
+			nnerrs = append(nnerrs, CacheProviderAccessError{css[i], err})
+		}
+	}
+	if len(nnerrs) > 0 {
+		if len(nnerrs) == 1 {
+			return nnerrs[0]
+		}
+		return MultipleErrors{nnerrs}
+	}
+	return nil
+}
+
+func (h *heptane) update(u Update) error {
+	tn := u.TableName
+	f := h.info(tn)
+	if f == nil {
+		return UnregisteredTableNameError{tn}
+	}
+	if f.RowProvider == nil {
+		return UnregisteredRowProviderError{tn}
+	}
+	key, err := f.Table.cacheKey(u.FieldValues)
+	if err != nil {
+		return err
+	}
+	ru := RowUpdate{f.Table, u.FieldValues}
+	if err := f.RowProvider.Access(ru); err != nil {
+		return RowProviderAccessError{ru, err}
+	}
+	if f.Table.isMissingSomeValue(u.FieldValues) {
+		rr := RowRetrieve{f.Table, u.FieldValues, nil}
+		if err := f.RowProvider.Access(&rr); err != nil {
+			return RowProviderAccessError{rr, err}
+		}
+		u.FieldValues = rr.RetrievedValues[0]
+	}
+	if f.CacheProvider == nil || f.Table.PrimaryKeyCachePrefix == nil {
+		return nil
+	}
+	value, err := f.Table.cacheValue(u.FieldValues)
+	if err != nil {
+		return err
+	}
+	cs := CacheSet{key.Key(), value.Value()}
+	if err := f.CacheProvider.Access(cs); err != nil {
+		return CacheProviderAccessError{cs, err}
+	}
+	return nil
+}
+
+func (h *heptane) delete(d Delete) error {
+	tn := d.TableName
+	f := h.info(tn)
+	if f == nil {
+		return UnregisteredTableNameError{tn}
+	}
+	if f.RowProvider == nil {
+		return UnregisteredRowProviderError{tn}
+	}
+	key, err := f.Table.cacheKey(d.FieldValues)
+	if err != nil {
+		return err
+	}
+	rd := RowDelete{f.Table, d.FieldValues}
+	if err := f.RowProvider.Access(rd); err != nil {
+		return RowProviderAccessError{rd, err}
+	}
+	if f.CacheProvider == nil || f.Table.PrimaryKeyCachePrefix == nil {
+		return nil
+	}
+	cs := CacheSet{key.Key(), make(cacheValue, 0).Value()}
+	if err := f.CacheProvider.Access(cs); err != nil {
+		return CacheProviderAccessError{cs, err}
+	}
+	return nil
+}
+
+func (h *heptane) Access(a Access) error {
+	switch a := a.(type) {
+	case Create:
+		return h.create(a)
+	case *Create:
+		return h.create(*a)
+	case *Retrieve:
+		return h.retrieve(a)
+	case Update:
+		return h.update(a)
+	case *Update:
+		return h.update(*a)
+	case Delete:
+		return h.delete(a)
+	case *Delete:
+		return h.delete(*a)
+	}
+	return UnsupportedAccessTypeError{a}
+}
+
+func (h *heptane) AccessSlice(aa []Access) (errs []error) {
```

Comment:
Is aa a proper parameter name?

Reply:
If you mean it is too short, https://www.reddit.com/r/golang/comments/8wxwgv/why_does_go_encourage_short_variable_names/.

If you mean a different thing, maybe you could suggest a more proper name?